File: core/src/main/java/no/elg/infiniteBootleg/console/HelpfulConsoleHelpUtil.java
package no.elg.infiniteBootleg.console;

import com.badlogic.gdx.utils.reflect.Annotation;
import com.badlogic.gdx.utils.reflect.ClassReflection;
import com.badlogic.gdx.utils.reflect.Method;
import com.strongjoshua.console.CommandExecutor;
import com.strongjoshua.console.Console;
import com.strongjoshua.console.ConsoleUtils;
import com.strongjoshua.console.annotation.ConsoleDoc;
import java.util.HashSet;
import java.util.Set;
import no.elg.infiniteBootleg.Main;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class HelpfulConsoleHelpUtil {

  public static void printHelp(Console console, CommandExecutor exec, String command) {

    Set<Method> methods = getRelevantMethods(console, exec, command);
    if (methods.isEmpty()) {
      Main.logger().log("Command does not exist.");
      return;
    }

    for (Method m : methods) {
      StringBuilder sb = createCmdPrefix(m);
      Annotation annotation = m.getDeclaredAnnotation(ConsoleDoc.class);
      if (annotation != null) {

        ConsoleDoc doc = annotation.getAnnotation(ConsoleDoc.class);
        sb.append(doc.description());

        Class<?>[] params = m.getParameterTypes();
        String[] names = getArgNames(m, params);

        for (int i = 0; i < params.length; i++) {
          sb.append("\n");
          for (int j = 0; j < m.getName().length() + 2; j++) {
            // using spaces this way works with monotype fonts
            sb.append(" ");
          }
          sb.append('<');
          sb.append(params[i].getSimpleName());
          if (names != null && i < names.length) {
            sb.append(' ');
            sb.append(names[i]);
          }
          sb.append("> ");
          if (i < doc.paramDescriptions().length) {
            sb.append(doc.paramDescriptions()[i]);
          }
        }
      } else {
        appendCmdSignature(sb, m);
      }

      Main.logger().log(sb.toString());
    }
  }

  public static void printCommands(Console console, CommandExecutor exec) {
    for (Method method : getRelevantMethods(console, exec, null)) {
      if (canExecute(method)) {
        StringBuilder sb = createCmdPrefix(method);
        appendCmdSignature(sb, method);
        Main.logger().log(sb.toString());
      }
    }
  }

  public static boolean canExecute(Method method) {
    return !Main.isServer() || !method.isAnnotationPresent(ClientsideOnly.class);
  }

  public static void appendCmdSignature(StringBuilder sb, Method method) {
    Class<?>[] params = method.getParameterTypes();
    String[] names = getArgNames(method, params);

    for (int i = 0; i < params.length; i++) {

      sb.append('<');
      sb.append(params[i].getSimpleName());

      if (names != null && i < names.length) {
        sb.append(' ');
        sb.append(names[i]);
      }
      sb.append('>');

      if (i < params.length - 1) {
        sb.append(' ');
      }
    }
  }

  public static String generateCommandSignature(Method method) {
    StringBuilder sb = createCmdPrefix(method);
    appendCmdSignature(sb, method);
    return sb.toString();
  }

  private static String[] getArgNames(Method method, Class<?>[] params) {
    Annotation annotation = method.getDeclaredAnnotation(CmdArgNames.class);
    String[] names = null;
    if (annotation != null) {
      names = annotation.getAnnotation(CmdArgNames.class).value();

      if (names.length != params.length) {
        Main.logger()
            .warn(
                "Command argument names annotation is present on command '%s', but there are too %s names. "
                    + "Expected %d "
                    + "names "
                    + "found %d",
                method.getName(),
                (names.length < params.length) ? "few" : "many",
                params.length,
                names.length);
      }
    }
    return names;
  }

  private static StringBuilder createCmdPrefix(Method method) {
    StringBuilder sb = new StringBuilder();
    sb.append(method.getName());
    sb.append(" ");
    return sb;
  }

  /**
   * @param console The console used
   * @param exec The executor to get the methods from, will look at all superclasses
   * @param command Command method to get, if null no filtering will be done
   * @return All relevant command method
   */
  public static Set<Method> getRelevantMethods(
      @NotNull Console console, @NotNull CommandExecutor exec, @Nullable String command) {
    Set<Method> methods = new HashSet<>();
    Class<?> c = exec.getClass();
    while (c != Object.class) {

      for (Method method : ClassReflection.getDeclaredMethods(c)) {
        if (method.isPublic()
            && //
            ConsoleUtils.canDisplayCommand(console, method)
            && //
            (command == null || method.getName().equalsIgnoreCase(command))) {

          methods.add(method);
        }
      }
      c = c.getSuperclass();
    }
    return methods;
  }
}

File: core/src/main/kotlin/no/elg/infiniteBootleg/console/HelpfulConsoleHelpUtil.kt
package no.elg.infiniteBootleg.console

import com.badlogic.gdx.utils.reflect.Annotation
import com.badlogic.gdx.utils.reflect.ClassReflection
import com.badlogic.gdx.utils.reflect.Method
import com.strongjoshua.console.CommandExecutor
import com.strongjoshua.console.Console
import com.strongjoshua.console.ConsoleUtils
import com.strongjoshua.console.annotation.ConsoleDoc
import io.github.oshai.kotlinlogging.KotlinLogging
import no.elg.infiniteBootleg.main.Main

private val logger = KotlinLogging.logger {}

object HelpfulConsoleHelpUtil {

  fun printHelp(console: Console, exec: CommandExecutor, command: String) {
    val methods = getRelevantMethods(console, exec, command)
    if (methods.isEmpty()) {
      logger.info { "Command does not exist." }
      return
    }
    for (m in methods.sortedBy(Method::name)) {
      val sb = createCmdPrefix(m)
      val annotation: Annotation? = m.getDeclaredAnnotation(ConsoleDoc::class.java)
      if (annotation != null) {
        val doc: ConsoleDoc = annotation.getAnnotation(ConsoleDoc::class.java) ?: continue
        sb.append(doc.description)
        val params = m.parameterTypes
        val names = getArgNames(m, params)
        for (i in params.indices) {
          sb.appendLine()
          // using spaces this way works with monotype fonts
          sb.append(" ".repeat((m.name.length + 2)))
          sb.cmdArgs(params, names, i)
          if (i < doc.paramDescriptions.size) {
            sb.append(doc.paramDescriptions[i])
          }
        }
      } else {
        sb.appendCmdSignature(m)
      }
      logger.info { sb.toString() }
      if (!allowedToExecute(m)) {
        logger.warn { "You are not allowed to execute command '" + m.name + "'" }
      }
    }
  }

  fun printCommands(console: Console, exec: CommandExecutor) {
    for (method in getRelevantMethods(console, exec, null).sortedBy(Method::name)) {
      if (allowedToExecute(method)) {
        val sb = createCmdPrefix(method)
        sb.appendCmdSignature(method)
        logger.info { sb.toString() }
      }
    }
  }

  fun allowedToExecute(method: Method): Boolean {
    val client = Main.isClient || !method.isAnnotationPresent(ClientsideOnly::class.java)
    val auth = Main.isAuthoritative || !method.isAnnotationPresent(AuthoritativeOnly::class.java)
    return client && auth
  }

  private fun StringBuilder.appendCmdSignature(method: Method): StringBuilder {
    val params = method.parameterTypes
    val names = getArgNames(method, params)
    for (i in params.indices) {
      this.cmdArgs(params, names, i)
    }
    return this
  }

  private fun StringBuilder.cmdArgs(params: Array<Class<*>>, names: Array<String>?, argNr: Int): StringBuilder {
    append('<').append(params[argNr].simpleName)
    if (names != null && argNr < names.size) {
      append(' ').append(names[argNr].replace('_', '-').replace(' ', '-'))
    }
    return append("> ")
  }

  fun generateCommandSignature(method: Method): String = createCmdPrefix(method).appendCmdSignature(method).toString()

  private fun getArgNames(method: Method, params: Array<Class<*>>): Array<String>? {
    val annotation = method.getDeclaredAnnotation(CmdArgNames::class.java)
    if (annotation != null) {
      val names: Array<out String> = annotation.getAnnotation(CmdArgNames::class.java).value
      if (names.size != params.size) {
        logger.warn {
          "Command argument names annotation is present on command '${method.name}', " +
            "but there are too ${if (names.size < params.size) "few" else "many"} names. Expected ${params.size} names found ${names.size}"
        }
      }
    }
    return null
  }

  private fun createCmdPrefix(method: Method): StringBuilder = StringBuilder().append(method.name).append(" ")

  private fun methodNameMatchesCommand(command: String?, method: Method): Boolean = command == null || method.name.startsWith(command, ignoreCase = true)

  /**
   * @param console The console used
   * @param exec The executor to get the methods from, will look at all superclasses
   * @param command Command method to get, if null no filtering will be done
   * @return All relevant command method
   */
  private fun getRelevantMethods(console: Console, exec: CommandExecutor, command: String?): Set<Method> {
    val methods: MutableSet<Method> = HashSet()
    var clazz: Class<*> = exec.javaClass
    while (clazz != Any::class.java) {
      for (method in ClassReflection.getDeclaredMethods(clazz)) {
        if (method.isPublic && ConsoleUtils.canDisplayCommand(console, method) && methodNameMatchesCommand(command, method)) {
          methods.add(method)
        }
      }
      clazz = clazz.superclass
    }
    if (methods.isEmpty()) {
      logger.warn { "Failed to find any relevant methods" }
    }
    return methods
  }
}
